### ISSUE TITLE
Failed to create a new stream through a proxy

### ISSUE THREAD

**Issue by ristowee** [LuatixHQ/opencti#4425]
## Description

The (ingestion/synchronization) stream creation fails if the connection needs to go through a proxy.

## Environment

1. OS (where OpenCTI server runs): Debian 10 (docker-compose)
2. OpenCTI version: 5.10.2
3. OpenCTI client: frontend
4. Other environment details: A squid proxy running on port 3128

## Reproducible Steps

1. Set up a proxy & two OpenCTI instances & serve a live stream on one instance
2. Configure the other instance to use the proxy to connect to the first instance (in my case I put the conf in the ~/.docker/config.json)
3. Open Data --> Ingestion --> add a new synchronizer with (+)
4. Fill in "Remote OpenCTI URL" and "Remote OpenCTI token" and press VALIDATE
5. Observe "Error getting the streams from remote OpenCTI" error in the screen

## Expected Output

Validation succeeds and I can select the stream I want to listen to.

## Actual Output

The validation fails with "Error getting the streams from remote OpenCTI" text.

## Additional information

It doesn't matter if the end point uses http or https. Both cases fail. The listener wants to connect to a wrong port. It tries to open the graphql path in the squid port (3128).

The validation is done differently from the actual streaming. It uses [`axios` calls](https://github.com/OpenCTI-Platform/opencti/blob/5.10.2/opencti-platform/opencti-graphql/src/domain/connector.js#L217) and the streaming uses [`EventSource`](https://github.com/OpenCTI-Platform/opencti/blob/5.10.2/opencti-platform/opencti-graphql/src/manager/syncManager.js#L45).

If a stream already exists it can be stopped and started without problems.

### Squid logs:

Failing HTTPS validation:
```
1695214005.291 180259 10.23.166.222 TCP_MISS/503 4366 POST https://10.23.166.215:3128/graphql - HIER_DIRECT/10.23.166.215 text/html
1695214005.291      0 10.23.166.222 NONE/000 0 NONE error:transaction-end-before-headers - HIER_NONE/- -
```

Failing HTTP validation:
```
1695373647.043  59320 10.23.166.222 TCP_MISS/503 4288 POST http://10.23.166.214:3128/graphql - HIER_DIRECT/10.23.166.214 text/html
1695373647.043      0 10.23.166.222 NONE/000 0 NONE error:transaction-end-before-headers - HIER_NONE/- -
```

Successful start for an existing stream:
```
1695214148.421     23 10.23.166.26 TCP_TUNNEL/200 5796 CONNECT 10.23.166.215:443 - HIER_DIRECT/10.23.166.215 -
```

## Screenshots (optional)

Wireshark capture from the "client" instance:
![opencti_proxy_problem](https://github.com/OpenCTI-Platform/opencti/assets/123449325/5da3ba24-fd58-4f2c-9c98-ed30d912ce0b)

**Comment by richard-julien** [comment]
Do you use environment variables to configure your proxy?

**Comment by ristowee** [comment]
Good point. If I set up the [~/.docker/config.json](https://docs.docker.com/network/proxy/) docker-compose automatically injects the following proxy variables in the env:
HTTP_PROXY
HTTPS_PROXY
NO_PROXY
http_proxy
https_proxy
no_proxy

**Comment by richard-julien** [comment]
Ok thanks. Do you have also this commit https://github.com/OpenCTI-Platform/opencti/commit/9b5e0749238faf5bebab11e95cae1bc5dd532ebb in your base code? If not can you try after applying it? 
thanks

**Comment by ristowee** [comment]
Yes, I do. I have also: https://github.com/OpenCTI-Platform/opencti/commit/de96ff647c8a07f4374c38a2a2d260c8f23b48cf, https://github.com/OpenCTI-Platform/opencti/commit/77f3f35062158a6ee9df2de20ff6bb878635dce7f, https://github.com/OpenCTI-Platform/opencti/commit/fbb5228df60da48520bc0b18ff10ab4bdaa5fd35. The proxying did not work at all without them.

**Comment by richard-julien** [comment]
Ok strange. We have exactly the same use case without any issue on our side. Any possibility to open your proxy and target platform to let me reproduce the problem?

**Comment by ristowee** [comment]
Bummer! Unfortunately, I cannot open you a connection to our systems. I'll try to figure out if I can reproduce it in a way I can share it.

**Comment by ristowee** [comment]
Hi @richard-julien,

I did a little testing and it seems something has broken in http-proxy-agent (and https-proxy-agent) between 5.0.0 and 6.0.1.

I can reproduce my problem with the following example (I basically copied the necessary pieces from opencti-platform codebase and pasted them in one file). I've omitted HTTPS because HTTP is easier to test. Lets call this file `proxy_test.js`:

```javascript
const nconf = require('nconf');
const axios = require('axios');
const HttpProxyAgent = require('http-proxy-agent').HttpProxyAgent;

nconf.env({ separator: '__', lowerCase: true, parseValues: true });

const getPlatformHttpProxies = () => {
  const http = nconf.get('http_proxy');
  const proxies = {};
  if (http) {
    proxies['http:'] = {
      build: () => new HttpProxyAgent(http),
      isExcluded: false
    };
  }
  return proxies;
};

const getHttpClient = ({ headers, responseType }) => {
  const proxies = getPlatformHttpProxies();
  return axios.create({
    responseType,
    headers,
    httpAgent: proxies['http:']?.build(),
  });
};

const callHttp = async() => {
  const proxies = getPlatformHttpProxies();
  // Test print to see what the proxy data looks like (different in 5.0.0 and newer)
  const agent = proxies["http:"].build();
  console.log(agent.proxy);

  const remoteUri = "http://httpbin.org/anything";
  const httpClientOptions = { responseType: 'json' };
  const httpClient = getHttpClient(httpClientOptions);
  const { data } = await httpClient.post(remoteUri, { 'foo': 'bar' });
  // Print return values
  console.log(data);
};

callHttp();
``` 

Install necessary packages: `npm install nconf axios http-proxy-agent`, install Squid, run test `http_proxy=http://localhost:3128 node proxy_test.js`.

With version 6.0.1 the `post`function takes forever and the Squid logs show:
```
1697107270.744   4732 ::1 TCP_MISS_ABORTED/000 0 POST http://httpbin.org:3128/anything - HIER_DIRECT/54.161.141.91 -
1697107270.744      0 ::1 NONE/000 0 NONE error:transaction-end-before-headers - HIER_NONE/- -
```
and the Wireshark confirms this:
![image](https://github.com/OpenCTI-Platform/opencti/assets/123449325/1c8d4032-40fa-4640-b259-1aa182f014e5)

If I use version 5.0.0 the logs look nicer:
```
1697107144.643    425 ::1 TCP_MISS/200 913 POST http://httpbin.org/anything - HIER_DIRECT/54.80.46.162 application/json
```
![image](https://github.com/OpenCTI-Platform/opencti/assets/123449325/401dd560-273c-41a4-a35b-c76fed08c607)

 So, basically the OpenCTI is not broken per se but the libraries it uses are not behaving well.

**Comment by richard-julien** [comment]
Hi @ristowee , can you take a look with "http-proxy-agent": "7.0.0" as is the current version in the master code?

**Comment by ristowee** [comment]
Yes. 7.0.0 was the one I checked first (got the version from OpenCTI yarn file) and it failed. I then tried to figure out the version and found out that 5.0.0 worked as expected but 6.0.1 didn't work anymore.
I'm trying to debug the http-proxy-agent but I'm still figuring out how to build it locally so I can add some prints. I can see that the output buffer in the `connect` function has the wrong port but I'm not sure yet, what the input `req` looks like and how it is manipulated.

**Comment by ristowee** [comment]
I was able to "fix" the http-proxy-agent by temporarily removing the port [80 override](https://github.com/TooTallNate/proxy-agents/blob/main/packages/http-proxy-agent/src/index.ts#L88). It erroneously checks the proxy port when it should be checking the target host port.

The https-proxy-agent is using wrong hosts as well but in a different way. If I replace [this line](https://github.com/TooTallNate/proxy-agents/blob/main/packages/https-proxy-agent/src/index.ts#L111) I can make the connection.
```diff
-               let payload = `CONNECT ${host}:${opts.port} HTTP/1.1\r\n`;
+               let payload = `CONNECT ${req.getHeader('host')} HTTP/1.1\r\n`;
```
However, there seems to be something fishy in axios. If I have a _https_ target host (e.g. https://httpbin.org/headers) and a _http_ proxy (e.g. http://localhost:3128) it uses the `httpAgent` and not the `httpsAgent`. So it starts the agent based on the proxy protocol and not the target URL protocol as one (and the http(s)-proxy-agent) would expect.

So, even with the https-proxy-connection change the proxy connection succeeds but then it tries to send unencrypted HTTP data to HTTPS port and that fails.

I'll probably have to check how the old versions of the agents work to try to figure out where the problems lies.

**Comment by ristowee** [comment]
Hi @richard-julien,

I think I found the reason why this didn't work. The problem was that even though we gave our custom the http agents to axios it was still using its own internal proxy system and the double proxying messed up the calls. Disabling the internal proxy fixed the situation for me:
```diff
diff --git a/opencti-platform/opencti-graphql/src/utils/http-client.ts b/opencti-platform/opencti-graphql/src/utils/http-client.ts
index 5fec6be55..0f0f62056 100644
--- a/opencti-platform/opencti-graphql/src/utils/http-client.ts
+++ b/opencti-platform/opencti-graphql/src/utils/http-client.ts
@@ -25,5 +25,6 @@ export const getHttpClient = ({ headers, rejectUnauthorized, responseType, certi
     headers,
     httpAgent: proxies['http:']?.build(),
     httpsAgent: proxies['https:']?.build() ?? defaultHttpsAgent,
+    proxy: false,
   });
 };
```
I got the idea from [this message](https://github.com/axios/axios/issues/3384#issuecomment-1172379435). There seems to be a lot of problems in axios proxy handling and people usually use separate agents to work around it.

**Comment by richard-julien** [comment]
As always thanks for the time spent to dig in this problem.
I have pushed the proxy false and also take care of using the same axios initalizer for all the remote calls done by the platform.
Will be released in 5.11
